rfcs/0116-artifact-service.md

```diff
@@ -0,0 +1,196 @@
+* Auth might not be done with scopes and tc-auth, but a shared secret between
+  Queue and Artifacts
+* The Queue will block new tasks after task resolution by not issuing new
+  signed urls
+* Data storage in Artifacts will be done with Postgres and not Azure Table
+  Storage
+* We do not intend to have an operation to list artifacts in the API
+* The redirecting logic of the Artifact service will adhere to the HTTP spec
+  using 300-series redirects where possible
+* Where content-negotiation is impossible, the service will block creation of
+  artifacts when incoming requests aren't compatible with required response
+* We will support uploading artifacts to different regions so that we can
+
+## API
+This API is up for debate and is only an illustrative starting point.  
+
+### Generalised request format
+There will be a concept of a generalised HTTP request.  This will be an object shaped like
+```javascript
+{
+  method: 'GET',
+  url: 'https://www.example.com',
+  headers: {
+    'user-agent': 'taskcluster',
+  }
+}
+```
+
+* The body will not be included, but can be described through HTTP headers in the request
+* Duplicated HTTP headers will be unsupported (i.e. key -> string not key -> list of strings)
+* All query string, path, fragements are contained in a fully formed `url` property
+
+### Creation
+
+```
+PUT /artifacts/:name
+```
+
+This endpoint will take an origin and name.  It will return a `200` with a
+document which contains a list of generalised requests which could be run to
+complete the actual upload.  Please see the queue's current `blob` storage type
+for an example of this pattern.  The body of the request will contain metadata
+about the artifact in the shape:
+
+```javascript
+{
+  origin: '10.10.10.10' || 's3_us-west-2',
+  contentType: 'application/json',
+  contentLength: 1234,
+  contentSha256: 'abcdef1234',
+  transferLength: 1234,
+  transferSha256: 'abcdef1234',
+  contentEncdoing: 
+  expiration: new Date(),
+}
+```
+
+The `origin` parameter will either be an IPv4 address, IPv6 address, or an
+identifier.  This parameter will specify the source location of the ultimately
+originating request.  This means the origin of the request to the Queue from
+the worker and not the request to the artifact service.  If the origin is an
+IPv4 or IPv6, a mapping of the address and IP will occur to find the backing
+storage.  If the origin is not an IPv4 or IPv6 address and is identical to a
+set of known and mappable identifiers (e.g. `s3_us-west-2`), then that will be
+used to map the token to find the backing storage.
+
+### Deletetion
+
+```
+DELETE /artifacts/:name
+DELETE /caches/artifacts/:name
+```
+
+Delete all copies of the named Artifact if using first endpoint.  The second
+endpoint would attempt delete all of the cached copies in the various
+locations.  This cache purging must return a `202 Accepted` response and not
+`200/204` due to the nature of how caches work in the real world.
```

Comment:
I'm confused what the difference is here between "copies" and "cached copies".

Reply:
"cached copy" is right now the thing that Cloud-Mirror creates.  It would also be the files in the CDN or a pull-through proxy.  The idea here is that we have a best-effort attempt to clear out whichever caches are stored for that named object.  This would be to support diagnostics from the Sheriffs or possibly even a service where after X failed attempts to fetch an artifact, we clear the caches and try 1 more time.  Maybe we have 100% reliable caches which never have errors and we don't need this ;)